### ISSUE TITLE
ICE: Support connecting to peer reflexive candidates

### ISSUE THREAD

**Issue by backkem** [pions/webrtc#335]
Support connecting to peer reflexive candidates.

cc @enobufs

**Comment by backkem** [comment]
I suggest we first verify we can connect to peer reflexive candidates correctly. Allow peer reflexive candidates to be added via `PeerConnection.AddIceCandidate()` and enable the ICE agent to connect to them. This may already work but it isn't handled correctly since `ice.CandidateTypePeerReflexive` is not fully implemented:
https://github.com/pions/webrtc/blob/361cef680f416a35570259222922e1b1336ebd65/pkg/ice/candidatetype.go#L10

Next, we can start generating our own peer reflexive candidates. For this case I think you are right. The peer reflexive candidates will only be generated once the ICE checks have started so they can only be passed via the `OnICECandidate()` callback. Internally the the logic for detecting peer reflexive candidates should be added to package `pkg/ice`.

**Comment by enobufs** [comment]
Attempted to connect from pion to Chrome, using pion's AddIceCandidate() for new candidates arrived after setRemoteDescription() was called, but ICE wouldn't succeed. (I confirmed, non-trickle mode in [my code](https://github.com/enobufs/webrtc-audio-streaming/tree/feature/pion) works).  I will look into this further.

**Comment by enobufs** [comment]
Ok, it was fixed by #350.  Confirmed trickle works.  Moving on to prflx issue...

**Comment by enobufs** [comment]
I noticed in the following configuration, pion ignores all the STUN-binding requests from NATed node. Consequently, all connectivity checks fail.
```
Chrome Browser ----> [NAT] ---((  Internet ))----> Pion
```

[This code](https://github.com/pions/webrtc/blob/20619e871258b242c19a8ed90dfbca6696492269/pkg/ice/agent.go#L650-L652) needs to be fixed.

Connectivity checks from a NAT'ed node will have translated source tranport address by NAT. Therefore, the receiving end won't be able to find the address in the remote candidate list (expected). It should respond to the request.  In the case for the receiving peer not being able to find the transport address in the current remote candidate list, it should be treated as a `peer-reflexive` candidate as described in -RFC 8445 - sec 7.3.1.3](https://tools.ietf.org/html/rfc8445#section-7.3.1.3)

**Comment by enobufs** [comment]
Added some text to our wiki page, to explain how  various candidates contribute to traversing NAT with a hope that helps: https://github.com/pions/webrtc/wiki/Network-Address-Translation#candidate-types-and-combinations-of-nat-types

**Comment by enobufs** [comment]
With the fix in PR #354, connection was successfully established with a prflx candidate.  Here's the log I am seeing:
```
LOG: 2019/01/12 13:11:08.558469 Signaling State has changed have-local-offer
ice DEBUG: 13:11:08.610730 agent.go:691: detected a new peer-reflexive candiate: 192.168.86.221:55342
ice DEBUG: 13:11:08.658742 agent.go:609: inbound isControlled && a.isControlling == false
ice DEBUG: 13:11:08.707453 agent.go:609: inbound isControlled && a.isControlling == false
LOG: 2019/01/12 13:11:08.710918 received remote description
LOG: 2019/01/12 13:11:08.711058 Signaling State has changed stable
ice DEBUG: 13:11:08.711115 agent.go:304: Started agent: isControlling? true, remoteUfrag: "UNdU", remotePwd: "1dusQw1EyD2hBKVmXHby9g8x"
LOG: 2019/01/12 13:11:08.711149 Connection State has changed Checking
ice DEBUG: 13:11:08.755911 agent.go:633: got controlling message: &stun.Message{Class:0x0, Method:0x1, Length:0x58, TransactionID:[]uint8{0x21, 0x12, 0xa4, 0x42, 0x38, 0x6c, 0x37, 0x4c, 0x71, 0x6b, 0x70, 0x78, 0x43, 0x76, 0x4c, 0x51}, Attributes:[]*stun.RawAttribute{(*stun.RawAttribute)(0xc0000a33e0), (*stun.RawAttribute)(0xc0000a3410), (*stun.RawAttribute)(0xc0000a3440), (*stun.RawAttribute)(0xc0000a3470), (*stun.RawAttribute)(0xc0000a34a0), (*stun.RawAttribute)(0xc0000a34d0)}, Raw:[]uint8{0x0, 0x1, 0x0, 0x58, 0x21, 0x12, 0xa4, 0x42, 0x38, 0x6c, 0x37, 0x4c, 0x71, 0x6b, 0x70, 0x78, 0x43, 0x76, 0x4c, 0x51, 0x0, 0x6, 0x0, 0x15, 0x62, 0x42, 0x6b, 0x4c, 0x41, 0x5a, 0x62, 0x56, 0x71, 0x50, 0x57, 0x56, 0x68, 0x52, 0x6b, 0x71, 0x3a, 0x55, 0x4e, 0x64, 0x55, 0x0, 0x0, 0x0, 0xc0, 0x57, 0x0, 0x4, 0x0, 0x0, 0x3, 0xe7, 0x80, 0x29, 0x0, 0x8, 0x91, 0xbe, 0xe1, 0x7a, 0x66, 0x64, 0x36, 0x87, 0x0, 0x24, 0x0, 0x4, 0x6e, 0x0, 0x1e, 0xff, 0x0, 0x8, 0x0, 0x14, 0xd8, 0x57, 0xf9, 0xc6, 0x3b, 0x62, 0xb6, 0x29, 0x1a, 0xac, 0xf3, 0xf6, 0x73, 0xc2, 0xa5, 0xf5, 0x58, 0x41, 0xc3, 0x35, 0x80, 0x28, 0x0, 0x4, 0x34, 0xd8, 0x26, 0x46}}
ice DEBUG: 13:11:08.755992 agent.go:376: Found valid candidate pair: host 192.168.86.221:54467 <-> prflx 192.168.86.221:55342 related :0 (selected? false)
ice DEBUG: 13:11:08.757339 agent.go:633: got controlling message: &stun.Message{Class:0x2, Method:0x1, Length:0x2c, TransactionID:[]uint8{0x21, 0x12, 0xa4, 0x42, 0x62, 0x61, 0x69, 0x43, 0x4d, 0x52, 0x41, 0x6a, 0x57, 0x77, 0x68, 0x54}, Attributes:[]*stun.RawAttribute{(*stun.RawAttribute)(0xc0001f2900), (*stun.RawAttribute)(0xc0001f2930), (*stun.RawAttribute)(0xc0001f2960)}, Raw:[]uint8{0x1, 0x1, 0x0, 0x2c, 0x21, 0x12, 0xa4, 0x42, 0x62, 0x61, 0x69, 0x43, 0x4d, 0x52, 0x41, 0x6a, 0x57, 0x77, 0x68, 0x54, 0x0, 0x20, 0x0, 0x8, 0x0, 0x1, 0xf5, 0xd1, 0xe1, 0xa8, 0xb4, 0x10, 0x2e, 0xa3, 0xda, 0x7b, 0xf8, 0x92, 0x8a, 0xcf, 0xee, 0x9f, 0x28, 0x3e, 0x4a, 0xfd, 0xaf, 0x2f, 0x57, 0x20, 0x85, 0x4a, 0x28, 0xa2, 0x4a, 0x85, 0x44, 0x9a, 0xb, 0x0, 0x0, 0x0, 0x18, 0xc0}}
ice DEBUG: 13:11:08.757410 agent.go:376: Found valid candidate pair: host 192.168.86.221:54467 <-> prflx 192.168.86.221:55342 related :0 (selected? true)
LOG: 2019/01/12 13:11:08.757490 Connection State has changed Connected
```
